src/eval/mod.rs

```diff
@@ -1227,9 +1228,12 @@ impl Eval for ast::Closure {
 }
 
 impl ast::Pattern {
-    // Destruct the given value into the pattern.
+    /// Destruct the given value into the pattern and apply the function to each binding.
     #[tracing::instrument(skip_all)]
-    pub fn define(&self, vm: &mut Vm, value: Value) -> SourceResult<Value> {
+    fn apply<T>(&self, vm: &mut Vm, value: Value, f: T) -> SourceResult<Value>
```

Comment:
the nesting is getting a bit excessive here, could this somehow be refactored / split up?

Reply:
Should be a lot more readable now. I'm not sure if I need to add any instrumentation macros, the `skip_all` makes me think I don't.